picamera2/picamera2.py

```diff
@@ -1141,14 +1145,20 @@ def process_requests(self) -> None:
                     self.functions = []
                     self._future.set_exception(e)
 
-            # Some applications may want to do something to the image after they've had a change
-            # to copy it, but before it goes to the video encoder.
-            if display_request and self.post_callback:
-                self.post_callback(display_request)
-
-            if self._encoder is not None:
+            if self.encode_stream_name in self.stream_map:
                 stream = self.stream_map[self.encode_stream_name]
-                self._encoder.encode(stream, display_request)
+
+            for req in requests:
+                # Some applications may want to do something to the image after they've had a change
+                # to copy it, but before it goes to the video encoder.
+                if self.post_callback:
+                    self.post_callback(req)
+
+                if self._encoder is not None:
+                    self._encoder.encode(stream, req)
+
+            for req in requests:
+                req.release()
```

Comment:
Maybe this could be the last line of the previous loop? (Sorry for the nit-pick!)

But otherwise the reference counting looks correct to me... :)

Reply:
Good point, just changed.